src/dispatch.rs

```diff
@@ -239,3 +252,35 @@ enum PutError {
     OpenTransactionError(OpenTransactionError),
     CommitError(db::CommitError),
 }
+
+async fn init_client_id(s: &dyn kv::Store) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
```

Comment:
The pattern I've been playing with for errors is to have a union that describes the potential errors for basically each function. Then if there is an internal error, I wrap the union value around that internal error (manually, there is no From implementation or boilerplate).

The result is basically a tree of strongly-typed errors that compactly describe every error that can happen in the entire call graph, which is pretty neat!

So for this, I'd say:

```
enum InitClientIDError {
    GetClientIDErr(kv::Error),
    OpenWriteTxErr(kv::Error),
    PutClientIDErr(kv::Error),
    CommitTxErr(kv::Error),
}
```

at the call sites, there's a pretty clean pattern:

```
let cid = s.get(CID_KEY).await.map_err(InitClientIDError::GetClientIDErr)?;
```

... or you can `use InitClientIDError::*` to make it shorter:

```
let cid = s.get(CID_KEY).await.map_err(GetClientIDErr)?;
```

I guess trying to extract some of the things I like about this:

- Every distinct error that can happen is strongly typed and visible to callers. You don't have to decide ahead of time which errors are important enough to give distinct values to make "catchable", and there's no separate machinery for that. And you can clearly see at each call site "what are the reasons this can fail?".
- All errors, no matter how obscure, get a relatively descriptive message by default
- You can easily map all errors to their source *and call graph* without line numbers
- Errors values are compact at runtime (just some integers) and stack allocated (I guess this doesn't matter though if errors are rare)

The downside is it that it's a bit burdensome to describe these enums. It would be nice to have a macro or some other machinery that generated them, but so far it's not that big a deal.

An alternative that would be less machinery would be the Go way, basically of prepending the context error message to the received error message and propagating it. But that loses the "what are the reasons this can fail?" transparency and also:

* There's no guarantee of string uniqueness so I always worry about tracking down the source of these errors
* There's mental overhead to wondering whether to wrap errors. If you don't wrap enough then it can be hard to figure out where an error came from. If you wrap too much the error messages get super long and hard to read.

Anyway I'm curious what you all think of adopting this pattern. So far I have enjoyed it. Try triggering an error in some of the code that uses it and looking at the resulting errors.

Anyway not a blocker for this, but we should discuss and narrow in on an error handling strategy that we use consistently throughout.